File: src/ax_eval/reporting.py
"""Render a run as a human-readable matrix + failure summary (markdown/text)."""

from __future__ import annotations

from collections import Counter


def _matrix_from_payload(payload: dict) -> tuple[list[str], list[str], dict]:
    harnesses = payload.get("harnesses", [])
    grid: dict[str, dict[str, bool]] = {}
    for r in payload.get("results", []):
        grid.setdefault(r["task_id"], {})[r["harness"]] = r["success"]
    task_ids = list(grid)
    return task_ids, harnesses, grid


def render(payload: dict) -> str:
    """Render a saved report payload (see storage.save_report) as text."""
    task_ids, harnesses, grid = _matrix_from_payload(payload)
    lines: list[str] = []
    title = f"AX eval — {payload.get('pack')} v{payload.get('pack_version')}"
    lines.append(title)
    lines.append("=" * len(title))
    lines.append("")

    if not task_ids:
        lines.append("(no results)")
        return "\n".join(lines)

    # Matrix table.
    tcol = max(len(t) for t in task_ids + ["task"])
    header = "task".ljust(tcol) + "  " + "  ".join(h.center(10) for h in harnesses)
    lines.append(header)
    lines.append("-" * len(header))
    for tid in task_ids:
        row = tid.ljust(tcol) + "  "
        cells = []
        for h in harnesses:
            ok = grid[tid].get(h)
            cells.append(("PASS" if ok else "FAIL").center(10))
        lines.append(row + "  ".join(cells))

    # Pass rates.
    lines.append("")
    lines.append("Pass rate by harness:")
    for h in harnesses:
        total = sum(1 for t in task_ids if h in grid[t])
        passed = sum(1 for t in task_ids if grid[t].get(h))
        pct = (passed / total * 100) if total else 0.0
        lines.append(f"  {h:<12} {passed}/{total}  ({pct:.0f}%)")

    # Top failing tasks (where most harnesses fail) — the "what to fix" hint.
    fails = Counter()
    for tid in task_ids:
        n_fail = sum(1 for h in harnesses if grid[tid].get(h) is False)
        if n_fail:
            fails[tid] = n_fail
    if fails:
        lines.append("")
        lines.append("Most-failed tasks (candidate fixes):")
        for tid, n in fails.most_common():
            lines.append(f"  {tid}: failed on {n}/{len(harnesses)} harnesses")

    return "\n".join(lines)

File: src/reporting.ts
/** Render a run as a human-readable matrix + failure summary (text/markdown). */
import { matrix, type RunReport } from "./runner.js";

export function render(report: RunReport): string {
  const harnesses = report.harnesses;
  // Reuse the canonical grid builder so the matrix and runner can't drift.
  // A cell is `undefined` when that (task, harness) pair was never run.
  const grid = matrix(report);
  const taskIds = Object.keys(grid);

  const lines: string[] = [];
  const title = `AX eval — ${report.pack} v${report.packVersion}`;
  lines.push(title, "=".repeat(title.length), "");

  if (taskIds.length === 0) {
    lines.push("(no results)");
    return lines.join("\n");
  }

  // A cell shows PASS/FAIL when run, or "—" when the pair was never attempted
  // (so the matrix can't disagree with the pass-rate denominator below).
  const cell = (tid: string, h: string): string => {
    const v = grid[tid]![h];
    return v === undefined ? "—" : v ? "PASS" : "FAIL";
  };

  // Matrix table.
  const tcol = Math.max(...taskIds.map((t) => t.length), "task".length);
  const header = "task".padEnd(tcol) + "  " + harnesses.map((h) => center(h, 10)).join("  ");
  lines.push(header, "-".repeat(header.length));
  for (const tid of taskIds) {
    const cells = harnesses.map((h) => center(cell(tid, h), 10));
    lines.push(tid.padEnd(tcol) + "  " + cells.join("  "));
  }

  // Pass rates — denominator is tasks actually attempted by that harness.
  lines.push("", "Pass rate by harness:");
  for (const h of harnesses) {
    const total = taskIds.filter((t) => grid[t]![h] !== undefined).length;
    const passed = taskIds.filter((t) => grid[t]![h] === true).length;
    const pct = total ? Math.round((passed / total) * 100) : 0;
    lines.push(`  ${h.padEnd(12)} ${passed}/${total}  (${pct}%)`);
  }

  // Most-failed tasks — the "what to fix" hint.
  const fails = taskIds
    .map((tid) => ({ tid, n: harnesses.filter((h) => grid[tid]![h] === false).length }))
    .filter((f) => f.n > 0)
    .sort((a, b) => b.n - a.n);
  if (fails.length) {
    lines.push("", "Most-failed tasks (candidate fixes):");
    for (const { tid, n } of fails) {
      lines.push(`  ${tid}: failed on ${n}/${harnesses.length} harnesses`);
    }
  }

  return lines.join("\n");
}

function center(s: string, width: number): string {
  if (s.length >= width) return s;
  const left = Math.floor((width - s.length) / 2);
  return " ".repeat(left) + s + " ".repeat(width - s.length - left);
}
